src/gs-update-monitor.c

```diff
@@ -150,6 +190,269 @@ no_updates_for_a_week (GsUpdateMonitor *monitor)
+	}
+}
+
+static void
+monitor_clear_scheduler (GsUpdateMonitor *monitor)
+{
+	g_clear_object (&monitor->scheduler);
+}
+
+static void
+async_result_cb (GObject *source_object,
+                 GAsyncResult *result,
+                 GAsyncResult **out_result)
+{
+	*out_result = g_object_ref (result);
+}
+
+static gboolean
+setup_scheduler (GsUpdateMonitor *monitor,
+		 GCancellable *cancellable,
+		 GError **error)
+{
+	g_autoptr(GAsyncResult) new_result = NULL;
+	GMainContext *context = NULL;
+	MwscScheduler *scheduler = NULL;
+
+	if (monitor->scheduler != NULL)
+		return TRUE;
+
+	context = g_main_context_default ();
+
+	mwsc_scheduler_new_async (cancellable,
+				  (GAsyncReadyCallback) async_result_cb,
+				  &new_result);
+	while (new_result == NULL)
+		g_main_context_iteration (context, TRUE);
+
+	scheduler = mwsc_scheduler_new_finish (new_result, error);
+
+	if (scheduler == NULL)
+		return FALSE;
+
+	g_signal_connect_swapped (scheduler, "invalidated",
+				  (GCallback) monitor_clear_scheduler,
+				  monitor);
+	monitor->scheduler = scheduler;
+
+	return TRUE;
+}
+
+static void
+scheduled_entry_invalidated_cb (MwscScheduleEntry *entry,
+				const GError *error,
+				UpdateScheduleHelper *helper)
+
+{
+	const gchar *app_id = gs_app_get_unique_id (helper->app);
+
+	g_debug ("Removing scheduled update of app %s", app_id);
+	g_hash_table_remove (helper->monitor->scheduled_updates, app_id);
+}
+
+static void
+schedule_update (GsUpdateMonitor *monitor,
+		 GsApp *app,
+		 GCancellable *cancellable)
+{
+	g_auto(GVariantDict) parameters_dict = G_VARIANT_DICT_INIT (NULL);
+	g_autoptr(GVariant) parameters = NULL;
+	g_autoptr(GError) local_error = NULL;
+	g_autoptr(GAsyncResult) new_result = NULL;
+	UpdateScheduleHelper *helper = NULL;
+	GMainContext *context = NULL;
+	const gchar *app_id = gs_app_get_unique_id (app);
+
+	helper = g_hash_table_lookup (monitor->scheduled_updates, app_id);
+	if (helper != NULL) {
+		/* replace the app that's scheduled, in case the object is
+		 * different */
+		if (helper->app != app) {
+			GCancellable *app_cancellable = gs_app_get_cancellable (helper->app);
+			g_cancellable_cancel (app_cancellable);
+			g_set_object (&helper->app, app);
+		}
+
+		return;
+	}
+
+	g_variant_dict_insert (&parameters_dict, "resumable", "b", FALSE);
+	parameters = g_variant_ref_sink (g_variant_dict_end (&parameters_dict));
+
+	if (!setup_scheduler (monitor, cancellable, &local_error)) {
+		g_warning ("Could not schedule update for app %s: %s",
+			   app_id, local_error->message);
+		return;
+	}
+
+	if (helper == NULL) {
+		helper = g_new0 (UpdateScheduleHelper, 1);
+		helper->monitor = monitor;
+		helper->app = g_object_ref (app);
+	}
+
+	mwsc_scheduler_schedule_async (monitor->scheduler, parameters, cancellable,
+				       (GAsyncReadyCallback) async_result_cb,
+				       &new_result);
+
+	context = g_main_context_default ();
+	while (new_result == NULL)
+		g_main_context_iteration (context, TRUE);
```

Comment:
This doesn't feel right: you're manually spinning the main loop of the global main context which will potentially block things while `mwsc_scheduler_schedule_async` is doing its work, right?

I think it would be better to either move the logic below into `async_result_cb` or, if you really want to do this thing of manually spinning the main loop, to make sure this is executed from a separate thread (and then probably use `g_main_context_get_thread_default` instead).

Reply:
If you’re manually iterating a main context, make sure it’s a new `GMainContext` instance specifically for this operation (and temporarily pushed onto the thread default stack). No need to run it from a separate thread.

Reply:
@mariospr See my response to the other similar comment.
@pwithnall I thought that if the code was running in the main thread it'd be fine like this, but if not, then I can do what you're saying. Maybe we could have this cumbersome synchronicity in a `..._sync` method in the Mogwai schedule client API.